src/rust/engine/workunit_store/src/lib.rs

```diff
@@ -570,15 +589,34 @@ impl WorkunitStore {
     let span_id = workunit.span_id;
     let new_metadata = Some(workunit.metadata.clone());
 
+    let workunit_counters = {
+      let mut counters = self.metrics_data.counters.lock();
+      match counters.entry(span_id) {
+        Entry::Vacant(_) => None,
+        Entry::Occupied(entry) => Some(entry.remove()),
+      }
+    };
```

Comment:
Any reason not to store these on the `Started` workunit rather than in a separate collection? I guess you'd need to wrap them each in a lock... but that's fairly cheap with `parking_lot` mutexes.

Reply:
There can be several work units that have not been completed, i.e. the current work unit and its parents. The collection tracks the increments on the work unit which was active at the time (keyed on the span ID). Counter increments that occurred in a parent while it was the only active work unit are only propagated on that parent work unit when it completes.

The work unit is stored on the stack and I'm not sure how I would access it safely from other code without wrapping it in some sort of smart pointer with locking. https://github.com/pantsbuild/pants/blob/9195e6023c428dfc80b7b1aef1e929fdd00625a6/src/rust/engine/workunit_store/src/lib.rs#L740

It seemed more straightforward to store the temporary state of counters on the `WorkunitStore` itself (via the `MetricsData` struct). That is already accessible in other task contexts. You will note that this pattern is already followed in how we send work units through channels. The data for the work unit is sent, not the work unit itself. So I just followed the existing pattern.

Reply:
At least that is true for the `Completed` message, which is where the final data for the work unit ends up.